src/Arcus.WebApi.Security/Authentication/SharedAccessKeyAuthenticationAttribute.cs

```diff
@@ -0,0 +1,104 @@
+        public SharedAccessKeyAuthenticationAttribute(string headerName, string secretName)
+        {
+            if (String.IsNullOrWhiteSpace(headerName))
+            {
+                throw new ArgumentException("Header name cannot be blank", nameof(headerName));
+            }
+
+            if (String.IsNullOrWhiteSpace(secretName))
+            {
+                throw new ArgumentException("Secret name cannot be blank", nameof(secretName));
+            }
+
+            _headerName = headerName;
+            _secretName = secretName;
+        }
+        
+        /// <inheritdoc />
+        public override async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (context.HttpContext?.Request == null)
+            {
+                throw new ArgumentException(
+                    "Invalid action context given without any HTTP request", 
+                    nameof(context.HttpContext.Request));
+            }
+
+            if (context.HttpContext.Request.Headers == null)
+            {
+                throw new ArgumentException(
+                    "Invalid action context given without any HTTP request headers",
+                    nameof(context.HttpContext.Request.Headers));
+            }
+
+            if (context.HttpContext.RequestServices == null)
+            {
+                throw new ArgumentException(
+                    "Invalid action context given without any HTTP request services", 
+                    nameof(context.HttpContext.RequestServices));
+            }
+
+            var userDefinedSecretProvider = context.HttpContext.RequestServices.GetService<ISecretProvider>();
+            if (userDefinedSecretProvider == null)
+            {
+                throw new KeyNotFoundException(
+                    $"No configured {nameof(ISecretProvider)} implementation found in the request service container. "
+                    + "Please configure such an implementation (ex. in the Startup) of your application");
+            }
+
+            var cachedSecretProvider = new CachedSecretProvider(userDefinedSecretProvider);
+            string foundSecret = await cachedSecretProvider.Get(_secretName);
+            if (foundSecret == null)
+            {
+                throw new AuthenticationException(
+                    $"No secret found with name {_secretName} in {nameof(ISecretProvider)} implementation {userDefinedSecretProvider.GetType().Name}");
+            }
+
+            bool noRequestSecretHeaderFound = 
+                context.HttpContext.Request.Headers
+                       .TryGetValue(_headerName, out StringValues requestSecretHeaders) == false;
+
+            bool anyRequestSecretHeaderDoesntMatch = 
+                requestSecretHeaders.Any(h => !String.Equals(h, foundSecret));
```

Comment:
what does `h` stand for? Pref to not have abbreviations

Reply:
Ah, ok. Didn't thought that this was part of the code style guide.
Usually with lambda expressions on sequences, we use the first letter of the type in the sequence.

So in this case it's a series of request headers, therefore the "H". But I'll rename it if you want.

Reply:
Personally I'm not a fan of this because it's not really clear what it represents but I'll leave it up to you!